keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1430 @@
+the (1.) case, and `(10+1)*2*10^4=0.21Mi` in the (2.) case. This means an increase
+of `0.19Mi`.
+
+The values of the limits are aligned with the values for the soft limits proposed
+as a fix for the for regular indexed jobs
+(see [here](https://github.com/kubernetes/kubernetes/issues/118085#issuecomment-1564520559)).
+However, in case when `backoffLimitPerIndex` is used we propose these limits
+to be hard.
+
+We believe that the scalability limits should be enough for most of Job use-cases.
+For workloads requiring larger jobs users should be able to create multiple Jobs,
+orchestrated by the [JobSet](https://github.com/kubernetes-sigs/jobset).
+
+### Expotential backoff delay issue
+
+Currently, a pod is recreated by the Job controller with expotential backoff
+delay (10s, 20s, 40s ...), counted from the last failure time.
+
+One complication is that the last failure time for failed pods may increase with
+time, as it fallbacks to `now` in some cases
+(see in [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/backoff_utils.go#L160-L182)).
+Thus, there is a risk that due to the presence
+of pods hitting the fallback the last failure time is continuously bumped,
+thus shifting the time to recreate the pod.
+
+This risk is present both when computing the expotential backoff delay globally
+(as for regular indexed Jobs), or per-index as proposed in in this KEP
+(see [Expotential backoff delay per index](#expotential-backoff-delay-per-index)).
+
+In order to mitigate this risk currently the time of last failure is recorded
+in-memory (globally for all pods within a Job). And a new failed pod may bump
+it only until it is added to the `uncountedTerminatedPods` structure.
+
+However, tracking the last failure time per index might be costly for memory
+consumption (see [Expotential backoff delay with in-memory tracking](#expotential-backoff-delay-with-in-memory-tracking)).
+
+Thus, in order to mitigate this risk we propose to compute the finish time for
+a pod as the first available value of the following (avoiding the ever-increasing
+fallback to `now`):
+1. max `finishAt` of all containers, if specified for all containers
+2. `LastTransitionTime` for the `Ready=False` condition
+3. `deletionTimestamp` - `deletionGracePeriodSeconds` if `deletionTimestamp` is set
+
+Here (3.) is used to mark the moment of deletion which is used to approximate
+the current behavior. (2.) is used when Kubelet loses track of one of its containers,
+the `Ready=False` condition is set by Kubelet when transitioning a pod to `Failed`
+phase: https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/kubelet/status/status_manager.go#L1060-L1068.
+When none of the above conditions is satisfied to compute the finish time we
+fallback to the pod's creation time.
+
+This fix can be considered a preparatory PR before the KEP, as to some extent
+is solves the preexisting issue.
+
```

Comment:
> Job controller maintains a queue of tasks (syncJob invocations). So events such as pod updates don't directly trigger syncJob, but put the owning job into the queue. The queue is processed by workers (default: 5) which are triggered every 1s. So the batching is by deduplication of items in the queue within the 1s intervals.

I think this may not be enough potentially and I would like an additional subsection to discuss this risk in more detail.
Based on what you wrote, in the current implementation we know that we're processing at most 1 job in a given moment, but in thery there is nothing that prevents us from starting processing it again immediately after that.
So let's say our "job processing handler" is very fast and finishes after 1ms. Then we can in theory process a job 1000 times per second, so if it has 1MB, we produce 1GB of data per second from it. That's not something that I want.

I guess what i would like to ensure is that we will do some additional batching here and ensure that we don't do Job updates (to update its status) more than X times per second or sth like that.

FTR: We implemented some batching mechanism in endpoint-slice and it helped a lot:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/endpointslice_controller.go#L484

Reply:
I might have simplified a bit.

The Job controller would not start a new `syncJob` just by time every 1s. The syncs are queued based on pod events (add, delete, update). The pod events enqueue `syncJob` with 1s delay, so if we had 1000 pod updates within 1s, they would all de-duplicate in the queue, and execute when 1s elapses. 

Here it is in code: https://github.com/kubernetes/kubernetes/blob/72a3990728b2a8979effb37b9800beb3117349f6/pkg/controller/job/job_controller.go#LL500C49-L500C69.

In fact the mechanism looks very similar to the one for endpointslices.

I'm not sure a new "risk" subsection is needed, because I don't see how the changes would affect the number or frequency of job updates. 

In particular, each pod termination (thus pod update, either success or failure) would trigger a delayed `syncJob`, resulting in Job status update, because the `status.failed` or `status.succeeded` fields are incremented. 

The difference is that before, only `status.completedIndexes` were updated for successful pods, and now also `status.failedIndexes` might be updated for failed pods. However, both fields are updated on the same Job status update as incrementing the `.status.failed` or `.status.succeeded` counters.

Reply:
I have added the ` Too fast Job status updates` section which describes the current queuing mechanism. 

However, now I'm thinking there might be a bug in the way the current mechanism is used. In particular, there is this line: https://github.com/kubernetes/kubernetes/blob/6195f96e56ee1e9f52986a0e768e22ca0d1949d6/pkg/controller/job/job_controller.go#L443 (which queues a new `syncJob` without a delay).  

Here, `pod_1` completion would result in Job status update, requeing a new `syncJob` without delay by line 443. Typically the next `syncJob` would be empty, so wouldn't requeue anything. However, if `pod_2` completes between the `syncJob`, then the invocation of `syncJob` will require another Job status update, thus requeuing another `syncJob` without delay (by the line 443). 

I think it is very unlikely that completing pods would create such a chain, so it is probably not a concern. However to be sure, we could adjust this line to also requeue with delay. Endpointslices do that. Is this something we are aware of, or it deserves an Issue? WDYT @alculquicondor @soltysh ?

Reply:
We have run load tests with different job sizes and they didn't reveal any problem https://github.com/kubernetes/perf-tests/tree/master/clusterloader2/testing/batch

But proactively, we could add a quick check: if `.metadata.generation` didn't change, requeue with delay.

Reply:
I have opened a PR for this issue: https://github.com/kubernetes/kubernetes/pull/118470.
However, it is a pre-existing issue in how we use the mechanism. So I suggest to track it outside of the KEP. wdyt @wojtek-t?

Reply:
:+1: for the linked PR, I left there one small nit about an integration or unit test which could simulate that case.

Reply:
yes - let's not track existing bug fixes as part of this KEP.